packages/core/src/types.ts

```diff
@@ -155,9 +155,25 @@ export type TransitionTargets<TContext> = Array<
   string | StateNode<TContext, any>
 >;
 
-export interface TransitionConfig<TContext, TEvent extends EventObject> {
-  cond?: Condition<TContext, TEvent>;
-  actions?: Actions<TContext, TEvent>;
+export type RestrictedActions<
+  TContext,
+  TEvent extends EventObject,
+  TEventType extends TEvent['type']
+> = SingleOrArray<
+  | ActionType
+  | AssignAction<TContext, TEvent & { type: TEventType }>
+  | AssignAction<TContext, TEvent>
```

Comment:
Shouldn't this part be this?
```suggestion
  | AssignAction<TContext, TEvent & { type: TEventType }>
```

Also - shouldn't `ActionObject` and `ActionFunction` be restricted here as well?